src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,245 @@
+        } catch (DataAccessException e) {
+            LOGGER.error("Following exception was thrown in addGiftCertificate(): " + e.getMessage());
+            throw new ServiceException("Failed to add certificate: certificate already exist",
+                    ErrorCodeEnum.FAILED_TO_ADD_CERTIFICATE);
+        } catch (PersistenceException e) {
+            LOGGER.error("Following exception was thrown: " + e.getMessage());
+            throw new ServiceException(e.getMessage(), ErrorCodeEnum.FAILED_TO_ADD_CERTIFICATE);
+        }
+    }
+
+    @Override
+    @Transactional
+    public void deleteGiftCertificate(int id) throws ServiceException {
+        GiftCertificate giftCertificate = getGiftCertificate(id);
+        if (giftCertificate == null) {
+            throw new ServiceException("Failed to delete certificate: certificate not found",
+                    ErrorCodeEnum.FAILED_TO_DELETE_CERTIFICATE);
+        }
+        try {
+            giftCertificateDAO.deleteAllCertificateTagRelations(giftCertificate.getId());
+
+            if (!giftCertificateDAO.deleteGiftCertificate(giftCertificate.getId())) {
+                LOGGER.error("Failed to delete certificate");
+                throw new ServiceException("Failed to delete certificate", ErrorCodeEnum.FAILED_TO_DELETE_CERTIFICATE);
+            }
+        } catch (DataAccessException e) {
+            LOGGER.error("Following exception was thrown in deleteGiftCertificate(): " + e.getMessage());
+            throw new ServiceException("Failed to delete certificate", ErrorCodeEnum.FAILED_TO_DELETE_CERTIFICATE);
+        }
+    }
+
+    @Override
+    @Transactional
+    public GiftCertificate updateGiftCertificate(GiftCertificate giftCertificate, int id) throws ServiceException {
+        giftCertificate.setId(id);
+        certificateValidator.validateCertificate(giftCertificate);
+        try {
+            giftCertificate.setLastUpdateDate(ZonedDateTime.ofInstant(Instant.now(), ZoneOffset.UTC));
+
+            addNewTagsToCertificate(giftCertificate);
+
+            if (!giftCertificateDAO.updateGiftCertificate(giftCertificate)) {
+                LOGGER.error("Failed to update certificate");
+                throw new ServiceException("Failed to update certificate", ErrorCodeEnum.FAILED_TO_UPDATE_CERTIFICATE);
+            }
+            return getGiftCertificate(id);
+        } catch (DataAccessException | PersistenceException e) {
+            LOGGER.error("Following exception was thrown in updateGiftCertificate(): " + e.getMessage());
+            throw new ServiceException("Failed to update certificate", ErrorCodeEnum.FAILED_TO_UPDATE_CERTIFICATE);
+        }
+    }
+
+    public void addNewTagsToCertificate(GiftCertificate giftCertificate) throws PersistenceException {
+        List<Tag> tagsInDataSource = tagDao.getAllTags();
+        for (Tag tag : giftCertificate.getTags()) {
+            if (!tagsInDataSource.contains(tag)) {
+                tag.setId(tagDao.addTag(tag));
```

Comment:
Why do you need to set id into the tag if you don`t return any value?

Reply:
because I need that id to create certificate tag relation, when tag is added